ARDOPC/Makefile

```diff
@@ -1,13 +1,72 @@
 #	ardopcf Makefile
+#		For Linux, the default build requires only gcc and libraries typically
+#		installed by default.
+#			cd ardop/ARDOPC
+#			make
+#
+#		Fow Windows, the default build requires installation of a MinGW build
+#		environment.  The easily installable packages from https://winlibs.com
+#		available for 32-bit or 64-bit builds are suggested.  These are used to
+#		build the Windows releases.  Installing these in `C:\Program Files` is
+#		not recommeneded since that may require admin privleges.  Other build
+#		environments may also work but are not tested.
+#			cd ardop\ARDOPC
+#			mingw32-make
+#
+#	`make test` which builds the executable and also runs some tests also
+#	requires installation of cmocka, which is not required for the default build.
+#		On Debian/Ubuntu this is easily installed with:
+#			sudo apt install libcmocka-dev
+#
+#		Package managers for other Linux distributions are also likely to
+#		provide easy installation of cmocka.
+#
+#		In the following description of how to install cmocka for Windows, a
+#		winlibs MinGW installation is assumed to be located at `C:\winlibs-32`
+#		or `C:\winlibs-64`.  If installed elsewhere, substitute the appropriate
+#		paths.  Putting the cmocka files into the winlibs install directory
+#		avoids the need for further configuration.
+#
+#		For Windows 32-bit:
+#			# Compiling from source as described for Windows 64-bit produces
+#			# errors, but precompiled files are available.
+#			download https://cmocka.org/files/1.1/cmocka-1.1.0-mingw.zip
+#			copy cmocka\bin\cmocka.dll from the zip to C:\winlibs-32\bin
+#			copy cmocka\include\*.h from the zip to C:\winlibs-32\include
+#			copy cmocka\lib\libcmocka.dll.a from the zip to c:\windlibs-32\lib
+#
+#		For Windows 64-bit:
+#			# Precompiled files like those used for Windows 32-bit do not appear
+#			# to be available.  So, cmocka will be compiled from source.
+#			git clone https://git.cryptomilk.org/projects/cmocka.git
+#			cd cmocka
+#			mkdir build
+#			cd build
+#			cmake -G "MinGW Makefiles" -DCMAKE_INSTALL_PREFIX="C:\winlibs=64" ..
+#			mingw32-make
+#			mingw32-make install
```

Comment:
I thought we were trying to avoid forcing the user to install cmake (or meson). Darn.

An important omission here is that the directory which contains `cmocka.dll` needs to be on the `PATH`. Otherwise it won't start. It also needs to be specified by absolute path in `gcc -L` unless that directory is already on the compiler's library search path.

Did you look at [vcpkg](https://vcpkg.link/ports/cmocka/v/1.1.5-1/0) by any chance? I've honestly not tried it. Does it install the libraries to the system path? Non-system libraries are a bit of a pain to support on both Windows and Linux both, but it is doable.

It would be easier if we could just build our own cmocka from source.

Reply:
I don't have anything specifically against cmake (or meson).  My primary interest is to make building ardopcf as simple as possible for people who are not serious programmers.  That includes minimizing the amount of stuff they have to install on their computers to build it.  On Linux, `make` is typically installed by default, while the others are not.

For Windows, no build tools are available by default.  However, the packages available from https://winlibs.com make it very easy to set up a gcc based build environment.  This install includes `cmake`.  So, using `cmake` to build cmocka for Windows requires no extra effort.

I'm also willing to accept a slightly more complex setup process for someone wanting to run the tests than for someone who just wants to be able to build ardopcf.  While not explicitly mentioned in the comments I added to the Makefile, I also used `git clone` to retrieve the cmocka source.  Thus, git must also be installed for this to work.  For a Windows user that wants to build ardopcf from source but doesn't want to run the tests, they can avoid the requirement for git by downloading the zipped source code for the desired branch from GitHub.

After unzipping a package from winlibs.com, the only setup needed to use it is to add its bin/ directory to the Windows PATH.  So, putting cmocka.dll in this directory ensures that it is in that path.  Similarly, by putting libcmocka.dll.a into the lib/ directory of the winlibs build system, it is where the build environment is already set to look for library files. 

I looked briefly at vcpkg.  However, everything I found for setting up vcpkg on Windows requires Visual Studio, which I'd rather avoid.

Reply:
> everything I found for setting up vcpkg on Windows requires Visual Studio, which I'd rather avoid

Ick. Agree 100%.

> I'm also willing to accept a slightly more complex setup process for someone wanting to run the tests

How about this: for a project of this size, it's not a big deal to maintain two build systems. We could make cmake a test dependency on all platforms. cmake can automate the entire build process for cmocka and run test-cases with ctest. If I go that route, would you prefer I vendor a copy of cmocka or retrieve source code at build time with [`FetchContent`](https://cmake.org/cmake/help/v3.14/module/FetchContent.html)? I really loathe build systems that reach out to the internet unnecessarily.

This would mean having to maintain both CMake and Makefile-based build rules, but this shouldn't be difficult. The Makefile would not need anything test-related.

Reply:
None of the machines I'm using to for development or testing of ardopcf are normally connected to the internet.  Other than occasional internet connections for updates and software installation, I only briefly connect one of them when I need to interact with GitHub.  So, anything that attempts to connect to the internet as a part of the build or testing process is unacceptable.  I find software that assumes it can always make an internet connection very annoying.  This is partly because I spend a lot of time in places where Internet connectivity is not available, but where I still want to continue working on this and other projects.

I know that I keep asking variants on this question, and I appreciate your patience, but what is the advantage of this dual build system?  Your example tests seemed to work well on Linux as you presented them, and with the changes I made to the Makefile, they worked on Windows.  So, this seemed like a good solution.  I haven't yet looked at the most recent versions of your testing branch (too busy writing comments), but I assume that it will also work well using only a single updated Makefile.  (See my comment above about my intention to further explore what when wrong when I tried to build cmocka with MinGW for 32-bit Windows)

Reply:
> This is partly because I spend a lot of time in places where Internet connectivity is not available

Vendoring it is.

> what is the advantage of this dual build system?

Consistency across platforms. With CMake, we can provide a very similar set of instructions for all users. CMake knows how to drive the compilers and the secondary build system. This makes it easier to support bash/gcc/make, bash/clang/ninja, powershell/msvc/msbuild, and whatever other bizarre combinations our devs want to use this week. Most of this works without any additional effort on our part—though sometimes you have to edit a line or two.

cmocka appears to be packaged for Debian, Ubuntu, Fedora, and Arch. This is unusually prolific for a library. But using the system library has some disadvantages:

1. We may need to support a wide range of cmocka versions. Thankfully their API is very stable.
2. If the packagers ship a buggy cmocka, we can't fix it. Again, not very likely.
3. If we ship a cmocka, we can build it without having to install something with root. (Or root alternatives like containers.) If we want to avoid Administrator on Windows, we should also strive to avoid root on Linux.

(3) is compelling, and it may eliminate a step for some devs.

Let me put together an alternative PR proposal that uses CMake building a vendored cmocka, in addition to the Makefile, and see if we like it. We can always "not like it" later.

Reply:
Since we agree that (1) and (2) are unlikely to be problems, it seems like a vendored cmocka may provide some advantage for Windows, but for Linux it just trades having to install (using sudo) cmocka for having to install (using sudo) cmake.  

Anybody using a less mainstream Linux distribution where cmocka isn't available through a package manager is presumably accustomed to figuring out how to install software themselves, probably including doing so from source.

FYI, since you mentioned it, I don't expect ardopcf to work with msvc, though I could be wrong.  Using getopt() is probably sufficient to eliminate this, and there may be other conflicts.

I tend to assume that most people using Linux have admin privileges, and are accustomed to using sudo when needed.  This is less likely for users of Windows machines since they are more likely to be used in schools, businesses, government offices, etc. (some of which may be used for amateur radio communications as a part of emergency operations plans) where users don't have full control.  Also, as I mentioned elsewhere, even with admin privileges on Windows, it can be inconvenient when using command line tools.

Reply:
As long as you're happy with your Windows installation process, and don't think it too burdensome, I'm inclined to leave it be for now and explore cmake-ification at a later time. Once the reorg proposed in #69 is complete, I will fold in your new Windows instructions and rebase this PR.